sound/soc/sof/ipc4-topology.c

```diff
@@ -0,0 +1,353 @@
+// SPDX-License-Identifier: (GPL-2.0-only OR BSD-3-Clause)
+//
+// This file is provided under a dual BSD/GPLv2 license.  When using or
+// redistributing this file, you may do so under either license.
+//
+// Copyright(c) 2021 Intel Corporation. All rights reserved.
+//
+//
+
+#include <linux/bits.h>
+#include <linux/device.h>
+#include <linux/errno.h>
+#include <linux/firmware.h>
+#include <linux/workqueue.h>
+#include <sound/pcm_params.h>
+#include <sound/sof/topology.h>
+#include <sound/tlv.h>
+#include <uapi/sound/sof/tokens.h>
+#include "ops.h"
+#include "sof-audio.h"
+#include "sof-priv.h"
+#include "ipc4.h"
+#include "ipc4-topology.h"
+#include "topology-tokens.h"
+
+/* 20th root of 10 in Q1.31 fixed-point notation*/
+#define VOL_FORTIETH_ROOT_OF_TEN	0x8f9e4d00ULL
+
+/* max value in Q1.31 fixed-point */
+#define IPC4_VOL_MAX	0x7fffffff
+/*
+ * Volume fractional word length define to 31 sets
+ * the volume linear gain value to use Q1.31 format
+ */
+#define VOLUME_FWL	31
+
+/* TLV data items */
+#define TLV_ITEMS	3
+#define TLV_MIN	0
+#define TLV_STEP	1
+#define TLV_MUTE	2
+
+static inline u32 vol_compute_64(u64 i, u32 x)
+{
+	return (u32)(((i >> (x - 1)) + 1) >> 1);
+}
+
+static inline int get_tlv_data(const int *p, int tlv[TLV_ITEMS])
+{
+	/* we only support dB scale TLV type at the moment */
+	if ((int)p[SNDRV_CTL_TLVO_TYPE] != SNDRV_CTL_TLVT_DB_SCALE)
+		return -EINVAL;
+
+	tlv[TLV_MIN] = (int)p[SNDRV_CTL_TLVO_DB_SCALE_MIN];
+
+	/* volume steps */
+	tlv[TLV_STEP] = (int)(p[SNDRV_CTL_TLVO_DB_SCALE_MUTE_AND_STEP] &
+				TLV_DB_SCALE_MASK);
+
+	/* mute ON/OFF */
+	if ((p[SNDRV_CTL_TLVO_DB_SCALE_MUTE_AND_STEP] &
+		TLV_DB_SCALE_MUTE) == 0)
+		tlv[TLV_MUTE] = 0;
+	else
+		tlv[TLV_MUTE] = 1;
+
+	return 0;
+}
+
+static int ipc4_set_up_volume_table(struct snd_sof_control *scontrol,
+				    int tlv[TLV_ITEMS], int size)
+{
+	u64 tmp;
+	int j;
+
+	/* init the volume table */
+	scontrol->volume_table = kcalloc(size, sizeof(u32), GFP_KERNEL);
+	if (!scontrol->volume_table)
+		return -ENOMEM;
+
+	scontrol->volume_table[0] = 0;
+	scontrol->volume_table[1] = tlv[TLV_MIN];
+	/* populate the volume table */
+	for (j = 2; j < size - 1; j++) {
+		tmp = (u64)scontrol->volume_table[j - 1] * VOL_FORTIETH_ROOT_OF_TEN;
+		scontrol->volume_table[j] = vol_compute_64(tmp, VOLUME_FWL);
+	}
+	scontrol->volume_table[size - 1] = IPC4_VOL_MAX;
+
+	return 0;
+}
+
+int sof_ipc4_control_load_volume(struct snd_soc_component *scomp,
```

Comment:
what is IPCv4 sepcific in this load_volume() compared to the IPC3 version?

Reply:
> what is IPCv4 sepcific in this load_volume() compared to the IPC3 version?

It is not about IPC, but about firmware. Different firmware can have their own volume definition. max, min, step, how to calculate, etc.